ipaserver/install/ipa_migrate.py

```diff
@@ -0,0 +1,2003 @@
+                                     stderr=subprocess.PIPE,
+                                     universal_newlines=True)
+            for stdout_line in iter(popen.stdout.readline, ""):
+                self.log_debug(stdout_line.rstrip())
+            for stdout_line in iter(popen.stderr.readline, ""):
+                self.log_debug(stdout_line.rstrip())
+            popen.stdout.close()
+            return_code = popen.wait()
+            if return_code:
+                self.log_error(f"ipa-server-upgrade failed: {return_code}")
+                self.post_notes.append("ipa-server-upgrade failed, "
+                                       "needs investigation")
+
+        # Run SIDGEN task
+        self.log_info("Running SIDGEN task ...")
+        if self.dryrun:
+            self.log_info("Skipping SIDGEN task in dryrun mode.")
+        else:
+            try:
+                cmd = ["/usr/bin/ipa config-mod --enable-sid --add-sids"]
+                result = subprocess.run(cmd, shell=True, check=True,
+                                        capture_output=True, text=True)
+                self.log_debug("SIDGEN task:\n" + result.stdout)
+            except subprocess.CalledProcessError as e:
+                self.log_error("SIDGEN task failed: " + str(e))
+                self.post_notes.append("SIDGEN task failed, "
+                                       "needs investigation.")
+
+        # TODO handle the LDIF conflict entries? (not used yet)
+
+        # Wrap it up with the summary report
+        self.display_stats(round(time.time() - start_time))
+
+    def run(self):
+        """
+        Run the IPA to IPA migration tool
+        """
+        # Validate user and setup
+        if not is_ipa_configured():
+            self.handle_error('IPA is not configured', err=2)
+
+        if os.getegid() != 0:
+            self.handle_error(f'Must be root to run {self.command_name}')
+
+        # Init the API
+        api.bootstrap(in_server=True, confdir=paths.ETC_IPA)
+        try:
+            api.finalize()
+        except Exception as e:
+            self.handle_error(f'Problem with IPA installation: {str(e)}',
+                              err=2)
+        api.Backend.ldap2.connect()
+        self.ldapiuri = realm_to_ldapi_uri(api.env.realm)
+        self.realm = api.env.realm
+        self.api = api
+        self.local_hostname = socket.getfqdn()
+        domain_parts = self.local_hostname.split(".")[1:]
+        self.local_domain = '.'.join(domain_parts)
+
+        # Check that we have kerberos credentials
+        try:
+            subprocess.run(["/usr/bin/ipa server-show "
+                            + self.local_hostname],
+                           capture_output=True,
+                           shell=True, check=True)
+        except subprocess.CalledProcessError:
+            self.handle_error("No kerberos credentials, please run "
+                              "'kinit admin' and try again")
```

Comment:
This is user-friendly but any member of the admins group can do the migration. Other tools, perhaps less friendly, report only "did not receive Kerberos credentials". IMHO that is preferable.